util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
+
+    detection_train_images = []
+    # creates a ScandirIterator aliased as files
+    print('------------ getting image names ------------')
+    with os.scandir(output_path) as files:
+      # loops through each file in the directory
+        for file in files:
+            if file.name.endswith('.jpg'):
+              # adds only the image files to the detection_train_imagess list
+                detection_train_images.append(file.name)
+   
+    # print('get bucket info')
+    # bucket = "calvinandpogs-ee148"
+    # detection_train_images_path = "atrw/detection/train"
+    # s3 = boto3.resource("s3")
+    # s3_bucket = s3.Bucket(bucket)
+    # print('bucket info collected')
+
+    # # this list holds all the image filename
+    # detection_train_images = [f.key.split(detection_train_images_path + "/")[1] for f in s3_bucket.objects.filter(Prefix=detection_train_images_path).all()]
+    print(str(len(detection_train_images)) + ' images collected')
+
+        
+    model = VGG16()
+    model = Model(inputs = model.inputs, outputs = model.layers[-2].output)
+
+    def extract_features(file, model):
+        # load the image from the s3 bucket as a 224x224 array and convert from 
+        # 'PIL.Image.Image' to numpy array because the VGG model expects the images
+        # it receives to be 224x224 NumPy arrays.
+
+        # object = s3.Object(bucket, detection_train_images_path + '/' + file)
+        # img = load_img(BytesIO(object.get()['Body'].read()), target_size=(224,224))
+        img = load_img(file, target_size=(224,224))
+
+        img = np.array(img) 
+        # reshape the data for the model reshape(num_of_samples, dim 1, dim 2, channels)
+        reshaped_img = img.reshape(1,224,224,3) 
+        # prepare image for model
+        imgx = preprocess_input(reshaped_img)
+        # get the feature vector
+        features = model.predict(imgx, use_multiprocessing=True)
+        return features
+    
+    data = {}
+    # p = r"CHANGE TO A LOCATION TO SAVE FEATURE VECTORS"
+
+    # loop through each image in the dataset
+    # temporarily trying first 1000 images
+    for i in tqdm(range(1000)):
+        # try to extract the features and update the dictionary
+        feat = extract_features(detection_train_images[i],model)
+        data[detection_train_images[i]] = feat
+        
+        # if something fails, save the extracted features as a pickle file (optional)
+        # except:
+        #     with open(p,'wb') as file:
+        #         pickle.dump(data,file)
+            
+    
+    # get a list of the filenames
+    filenames = np.array(list(data.keys()))
+
+    # get a list of just the features
+    feat = np.array(list(data.values()))
+
+    # reshape so that there are 210 samples of 4096 vectors
+    feat = feat.reshape(-1,4096)
```

Comment:
Is 4096 the output dim of VGNet? What are the 210 samples, I thought there were 1000?

The comment doesn't really describe what is happening here. Only talk about what the code is doing generically (e.g. "split into (samples, features)"), not for a specific run on specific inputs (although I still don't know where 210 came from).

Reply:
I believe 4096 is output dim of VGNet but not entirely sure. 
Sorry I should have removed the line mentioning 210 samples. That is the size of the dataset they used in the blog post

Reply:
Here's the VGGNet architecture:
![](https://neurohive.io/wp-content/uploads/2018/11/Capture-564x570.jpg)